### ISSUE TITLE
FP16 weights out of range error

### ISSUE THREAD

**Issue by rtrobin** [NVIDIA/TensorRT#420]
## Description

TRT7 onnx parser error when building fp16 engine

> Misc Error in operator(): 1 (Weights are outside of fp16 range. A possible fix is to retrain the model with regularization to bring the magnitude of the weights down.)

## Environment

TensorRT Version: 7.0.0.11
GPU Type: v100
Nvidia Driver Version: 440.33.01
CUDA Version: 10.2
CUDNN Version: 7.6.5
Operating System + Version: Ubuntu 18.04
Python Version (if applicable): 3.6.9
PyTorch Version (if applicable): 1.4.0
Baremetal or Container (if container which image + tag): ngc docker tensorrt:20.01-py3

## Steps To Reproduce

I have an onnx model converted from pytorch code, which could be built in fp32 mode successfully by onnx parser. However it failed when building args was set to fp16 mode. The error message is shown upon.

The question is, how to locate the specified layer that is overflowed? Otherwise, how to retrain the model to ensure the weights won't overflow? Using AMP?

Thanks in advance.

**Comment by rmccorm4** [comment]
If you set the tensorrt logger to verbose, it might show the specific layer, not sure though:

```
trtexec --explicitBatch --onnx=model.onnx --verbose
```

I don't know if AMP will "definitely ensure" no overflow, but it's definitely a good practice to use AMP as it should generally better utilize tensor cores (on applicable GPUs like your V100) and speed up training time.

**Comment by rtrobin** [comment]
It is a very long verbose message, but I can't see anything about layer with precision is shown.

> [E] [TRT] ../builder/cudnnBuilderWeightConverters.cpp (555) - Misc Error in operator(): 1 (Weights are outside of fp16 range. A possible fix is to retrain the model with regularization 
to bring the magnitude of the weights down.)

I know how AMP would help on training phase. But for the inference phase, is there any 'best practise' or 'programming guide' that indicates how to regularization fp16 weights, like int8 calibration in docs?

**Comment by rmccorm4** [comment]
> is there any 'best practise' or 'programming guide' that indicates how to regularization fp16 weights, like int8 calibration in docs?

That's a good question. I'll try to look into this tomorrow.

**Comment by rmccorm4** [comment]
> It is a very long verbose message, but I can't see anything about layer with precision is shown.

You can redirect the outputs to a file and upload here (drag and drop on comment), and I'll try to take a look at that tomorrow as well.

**Comment by rtrobin** [comment]
I will run the script and upload it later. I will try to use a smaller model to demonstrate this issue. Thank you in advance.

**Comment by rtrobin** [comment]
Please check the log file attached below. 
[fp16.log](https://github.com/NVIDIA/TensorRT/files/4304884/fp16.log)

**Comment by xiaochus** [comment]
Same problem. The onnx model is ok on TensorRT 5.1.5 but failure on TensorRT 7.0 C++ API。

**Comment by rmccorm4** [comment]
Hi @xiaochus,

You're getting the same error of weights being out of fp16 range? Or a different error?

The weights are saved in the ONNX model, so I don't see why it would work in TensorRT5, but not TensorRT7, because the weights are the same either way.

**Comment by rmccorm4** [comment]
Also @rtrobin, sorry I haven't got a response yet on some of your open questions, I'll try to get some answers.

**Comment by xiaochus** [comment]
@rmccorm4 Yes, same fp16 range error as below. I am trying to update TRT 5.1.5 to TRT 7.0 in our project, all onnx model were run well in both fp16 and int8 mode on TRT 5.1.5 but failure on TRT 7.0. I am trying to update pytorch to 1.3 and re-genrate onnx file, but it do not work. 

```
----------------------------------------------------------------
Input filename:   /model/erfnet.onnx
ONNX IR version:  0.0.3
Opset version:    9
Producer name:    pytorch
Producer version: 0.4
Domain:           
Model version:    0
Doc string:       
----------------------------------------------------------------
WARNING: onnx2trt_utils.cpp:198: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.

......

WARNING: onnx2trt_utils.cpp:198: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
setFp16Mode
ERROR: ../builder/cudnnBuilderWeightConverters.cpp (555) - Misc Error in operator(): 1 (Weights are outside of fp16 range. A possible fix is to retrain the model with regularization to bring the magnitude of the weights down.)
ERROR: ../builder/cudnnBuilderWeightConverters.cpp (555) - Misc Error in operator(): 1 (Weights are outside of fp16 range. A possible fix is to retrain the model with regularization to bring the magnitude of the weights down.)
```

**Comment by rmccorm4** [comment]
@xiaochus can you share your ONNX model so I can further look into it?

**Comment by xiaochus** [comment]
@rmccorm4 Sorry, i can't share the onnx file because they are on our company's server.

**Comment by rmccorm4** [comment]
@rtrobin unfortunately there isn't currently a good way to debug which op had fp16 weights that were out of range. The only workaround I can think of is to use the `setPrecision` on each of the potential fp16 layers along with the `STRICT_TYPES` flag to search for the bad layer(s). Once you've found the bad layer(s), then manually set their precisions to FP32, something like `badLayer->setPrecision(kFLOAT)`, so that you can still get the perf gain from the other FP16 layers within range.

I've filed an RFE to try to make this better.

**Comment by rmccorm4** [comment]
> is there any 'best practice' or 'programming guide' that indicates how to regularization fp16 weights, like int8 calibration in docs?

We don't have any recommendations on how to regularize fp16 weights to fit within range after training.

**Comment by rtrobin** [comment]
> The only workaround I can think of is to use the `setPrecision` on each of the potential fp16 layers along with the `STRICT_TYPES` flag to search for the bad layer(s). 

Hi, @rmccorm4 . Are you referring to build tensorrt engine by network api? In my case, I'm parsing the model by onnx parser. Or, maybe you talking about modifying onnx layer properties? Is there any api doc for onnx layer methods in this case?

**Comment by rmccorm4** [comment]
@rtrobin using the network API on the network object after parsing ONNX model

**Comment by rtrobin** [comment]
@rmccorm4 Ok, I see. Thanks for your advice.

**Comment by jawenzhang** [comment]
@rtrobin did you have resolved this problem?

**Comment by rtrobin** [comment]
@jawenzhang Unfortunately, no. But for my case, I cut the model weights to half in pyTorch code, and then convert it into trt fp16 engine. The accuracy only drops 0.1%, which is acceptable to me.

But of course, the advice above from @rmccorm4 is the most elegant way to convert the engine without accuracy lost and benefits from tensor core speedup. Though it needs much more efforts to put into.

**Comment by RicardSilva** [comment]
I came across this exact error. It looks like if you obtain the INetworkDefinition using a parser and the parser gets destroyed then the Engine builds fails. 

But if you keep the parser in scope after parsing the network then the engine building succeeds...

**Comment by rajeevsrao** [comment]
> I came across this exact error. It looks like if you obtain the INetworkDefinition using a parser and the parser gets destroyed then the Engine builds fails.
> 
> But if you keep the parser in scope after parsing the network then the engine building succeeds...

The INetworkDefinition object should be owned by the application and not the parser object. Can you please give an example of this bug if you have one @RicardSilva. Thanks.

**Comment by rajeevsrao** [comment]
I'm closing this issue. In conclusion, the model weights can be converted to half prior to conversion from torch and to identify any ops with weights out of precision range, do a binary search of candidate layers by forcing precision with setPrecision and strict type as @rmccorm4 suggested earlier in the thread. @rmccorm4 is there an RFE against onnx-tensorrt or TensorRT?

**Comment by rmccorm4** [comment]
Hi @rajeevsrao,

> @rmccorm4 is there an RFE against onnx-tensorrt or TensorRT?

Yes, I'll ping you.

**Comment by RicardSilva** [comment]
> > I came across this exact error. It looks like if you obtain the INetworkDefinition using a parser and the parser gets destroyed then the Engine builds fails.
> > But if you keep the parser in scope after parsing the network then the engine building succeeds...
> 
> The INetworkDefinition object should be owned by the application and not the parser object. Can you please give an example of this bug if you have one @RicardSilva. Thanks.

I don't have code available atm but what triggered this error for me:
1. Obtain a INetworkDefinition using some parser (ONNX parser in my case)
2. Destroy the Parser object
3. Try to build an engine using the network definition

If the order of 2 and 3 are swapped then the engine construction succeeds without any problem. It is hard for me to say what could cause this but it looks like there is some unintended dependency between the INetworkDefinition and the parser.

**Comment by xiaoxiao42** [comment]
> 

exactly 
you saved my life
